src/core/reducers/addons.js

```diff
@@ -261,10 +261,53 @@ export function createInternalAddon(
   return addon;
 }
 
-const initialState = {};
+type AddonID = number;
+
+export type AddonState = {|
+  // Flow wants hash maps with string keys.
+  // See: https://zhenyong.github.io/flowtype/docs/objects.html#objects-as-maps
+  byID: { [addonId: string]: AddonType },
+  byGUID: { [addonGUID: string]: AddonID },
+  bySlug: { [addonSlug: string]: AddonID },
+|};
+
+export const initialState: AddonState = {
+  byID: {},
+  byGUID: {},
+  bySlug: {},
+};
+
+export const getAddonByID = (
+  state: { addons: AddonState }, id: AddonID
+): AddonType | null => {
+  return state.addons.byID[`${id}`] || null;
+};
+
+export const getAddonBySlug = (
+  state: { addons: AddonState }, slug: string
+): AddonType | null => {
+  const addonId = state.addons.bySlug[slug];
+
+  return getAddonByID(state, addonId);
+};
+
+export const getAddonByGUID = (
+  state: { addons: AddonState }, guid: string
+): AddonType | null => {
+  const addonId = state.addons.byGUID[guid];
 
-export type AddonState = {
-  [addonSlug: string]: AddonType,
+  return getAddonByID(state, addonId);
+};
+
+export const getAllAddons = (
+  state: { addons: AddonState }
+): Array<AddonType> => {
+  const addons = state.addons.byID;
+
+  // TODO: one day, Flow will get `Object.values()` right but for now... we
+  // have to deal with it.
+  // See: https://github.com/facebook/flow/issues/2221.
+  return Object.keys(addons).map((key: string) => addons[key]);
```

Comment:
Can't we just tell flow to ignore this line instead? This seems silly (and while it might not be a huge deal, possibly less efficient in case `Object.values()` is optimised by the browser better than this inline function).

Reply:
I am fine with ignoring this line but I think it will confuse Flow for the rest of the code. Since I did that for the `languageTools` reducer, I'd like to follow-up in another patch. WDYT?

Reply:
What happens if you do `// $FlowFixMe` on that one line? If it doesn't mess stuff up I'd prefer our code not be made more onerous because of Flow.

Reply:
Alright doing this now. I'll patch the `languageTools` reducer too then (in a separate PR). I think we do that many times though...

Reply:
Oh you're right, but I realised that and it started to really bug me. A tool for writing good code shouldn't make us write worse code 😉

Reply:
I could not agree more.